src/Language/Futhark/Unused.hs

```diff
@@ -0,0 +1,207 @@
+module Language.Futhark.Unused (findUnused, getBody, getDecs) where
+
+import Data.Bifunctor qualified as BI
+import Data.Foldable (Foldable (foldl'))
+import Data.List.NonEmpty qualified as NE
+import Data.Map.Strict qualified as M
+import Data.Maybe (catMaybes, maybeToList)
+import Data.Set qualified as S
+import Language.Futhark
+import Language.Futhark.Semantic (FileModule (FileModule), includeToFilePath)
+
+-- Steps:`
+-- ∘ For each definition of a VName x in any file (corresponding to a top level definition), find which VNames are mentioned in the definition of x. (Intuitively, which functions are called directly.)
+-- ∘ Compute the transitive closure of the information gathered in (1), which means you now have a mapping from each VName to the set of VNames used in its definition (even indirectly).
+-- ∘ For those VNames defined in root files, take the union of their reachable names.
+-- ∘ Subtract the set produced in (3) from the set of all VNames. This gives you a set of VNames that are not reachable from the root files.
+
+-- Guidelines:
+-- (1) create a mapping for functions in *any* import to the directly usable functions in it.
+-- (2) create a mapping from normalized filename to top-level functions inside the filename.
+-- (3) Compute transitive closure for the map in (1)
+
+-- We are looking for VNames that represent functions.
+
+-- cabal run futhark -- unused un/b.fut
+
+getDecs :: FileModule -> [DecBase Info VName]
+getDecs (FileModule _ _env (Prog _doc decs) _) = decs
+
+findUnused :: [FilePath] -> [(ImportName, FileModule)] -> M.Map FilePath [(VName, SrcLoc)]
+findUnused fp fml = do
+  let fml' = map (BI.first includeToFilePath) fml
+      bf = M.unions $ map (funcsInFMod . snd) $ filter (\(x, _) -> x `elem` fp) fml'
+      rms = filter (\(x, _) -> x `notElem` fp) fml'
+      rf = map (BI.second funcsInFMod) rms
+      locs = M.unions $ map (locsInFMod . snd) rms
+      used1 = tClosure bf $ M.union bf $ M.unions $ map snd rf
+      used = S.unions $ map snd $ M.toList used1
+  M.fromList $
+    map
+      ( \(x, y) ->
+          ( x,
+            map (\a -> (a, locs M.! a)) $
+              filter (`M.member` locs) $
+                S.toList $
+                  S.fromList (map fst $ M.toList y) `S.difference` used
+          )
+      )
+      rf
+
+type VMap = M.Map VName (S.Set VName)
+
+type LocMap = M.Map VName SrcLoc
+
+-- possible future optimization:  remove VNames that aren't referenced in any top level import
+tClosure :: VMap -> VMap -> VMap
+tClosure bf af =
+  let bf' = tStep af bf
+   in if bf == bf'
+        then bf
+        else tClosure bf' af
+
+tStep :: VMap -> VMap -> VMap
+tStep af = M.map (\x -> S.union x $ S.unions $ S.foldl (<>) [] $ S.map (\y -> maybeToList $ y `M.lookup` af) x)
```

Comment:
This can be written more elegantly.  The last part certainly as something like:

    mapMaybe (`M.lookup` af)

Reply:
i have rewritten this with mapMaybe.